### PR TITLE
Feat/gho manager role t6247

### PR DESCRIPTION
**Pull request by foodaka**
Adds the gho manager contract which has the responsibility to update the gho discount strategy, discount lock period, and interest rate strategy for gho.

### CODE REVIEW

**Comment by foodaka** [comment]
Link-T6247

**Comment by foodaka** [comment]
Link T-6247

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * @notice GhoManager contract for setting Gho interest rate strategy, discount rate strategy, and discount lock period
+ */
+contract GhoManager {
+  // The Access Control List manager contract
+  IACLManager private immutable ACL_MANAGER;
+
+  /**
+   * @dev Only risk or pool admin can call functions marked by this modifier.
+   */
+  modifier onlyGhoManager() {
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:24]: This modifier is not needed. Risk/Pool admins are already able to call these functions.
We need to make this contract Ownable, so only the owner can call is functions

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /**
+   * @notice Updates the ReserveInterestRateStrategy for Gho
+   * @param _poolConfigurator The address of PoolConfigurator contract
+   * @param asset The address of the GHO deployed contract
+   * @param newRateStrategyAddress The address of new RateStrategyAddress contract
+   */
+  function setReserveInterestRateStrategyAddress(
+    PoolConfigurator _poolConfigurator,
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:61]: We can also fix the poolConfigurator address so we do not need to pass its address here. No strong opinion.

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    transferOwnership(owner);
+  }
+
+  /**
+   * @notice Updates the Discount Rate Strategy
+   * @param _ghoVariableDebtToken The address of GhoVariableDebtToken contract
+   * @param newDiscountRateStrategy The address of DiscountRateStrategy contract
+   */
+  function updateDiscountRateStrategy(
+    GhoVariableDebtToken _ghoVariableDebtToken,
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:36]: Better to use interfaces, whenever is possible. Same with the PoolConfigurator.
We can also be more flexible and pass addresses as params (we can cast later)

File: src/test/gho-manager.test.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  before(async () => {
+    ethers = hre.ethers;
+
+    const { pool } = testEnv;
+
+    poolSigner = await impersonateAccountHardhat(pool.address);
+    randomSigner = await impersonateAccountHardhat(ONE_ADDRESS);
+    poolConfigurator = await getPoolConfiguratorProxy();
+  });
+
```
Comment (miguelmtzinf) [inline src/test/gho-manager.test.ts:30]: you can also add a check to ensure GhoManager is PoolAdmin

Comment (foodaka) [reply src/test/gho-manager.test.ts:30]: good one `7041999aa44238888840cdc49bf17980fdcc3e8d`

File: src/contracts/facilitators/aave/misc/GhoManager.sol (hunk 3 above)
Comment (foodaka) [reply src/contracts/facilitators/aave/misc/GhoManager.sol]: updated to use interfaces

File: src/contracts/facilitators/aave/misc/GhoManager.sol (hunk 2 above)
Comment (foodaka) [reply src/contracts/facilitators/aave/misc/GhoManager.sol]: hmm can do -- do i hardcode this address?

File: src/contracts/facilitators/aave/misc/GhoManager.sol (hunk 1 above)
Comment (foodaka) [reply src/contracts/facilitators/aave/misc/GhoManager.sol]: removed

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+// Aave Contracts
+import {IACLManager} from '@aave/core-v3/contracts/interfaces/IACLManager.sol';
+import {IPoolAddressesProvider} from '@aave/core-v3/contracts/interfaces/IPoolAddressesProvider.sol';
+import {Errors} from '@aave/core-v3/contracts/protocol/libraries/helpers/Errors.sol';
```
[the remark this answers is not in the document]
Comment (foodaka) [reply src/contracts/facilitators/aave/misc/GhoManager.sol]: removed

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import {IACLManager} from '@aave/core-v3/contracts/interfaces/IACLManager.sol';
+import {IPoolAddressesProvider} from '@aave/core-v3/contracts/interfaces/IPoolAddressesProvider.sol';
+import {IPoolConfigurator} from '@aave/core-v3/contracts/interfaces/IPoolConfigurator.sol';
+import {IGhoVariableDebtToken} from 'src/contracts/facilitators/aave/tokens/interfaces/IGhoVariableDebtToken.sol';
+
+// OZ Contracts
+import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
+
+// Gho Contracts
+import {GhoVariableDebtToken} from 'src/contracts/facilitators/aave/tokens/GhoVariableDebtToken.sol';
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:14]:
```suggestion
import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {IPoolConfigurator} from '@aave/core-v3/contracts/interfaces/IPoolConfigurator.sol';
import {IGhoVariableDebtToken} from 'src/contracts/facilitators/aave/tokens/interfaces/IGhoVariableDebtToken.sol';
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * @title GhoManager
+ * @author Aave
+ * @notice GhoManager contract for setting Gho interest rate strategy, discount rate strategy, and discount lock period
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:19]:
```suggestion
 * @notice Helper contract for managing key risk parameters of the GHO reserve within the Aave Facilitator
 * @dev This contract is intended to be granted as PoolAdmin
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    transferOwnership(owner);
+  }
+
+  /**
+   * @notice Updates the Discount Rate Strategy
+   * @param _ghoVariableDebtToken The address of GhoVariableDebtToken contract
+   * @param newDiscountRateStrategy The address of DiscountRateStrategy contract
+   */
+  function updateDiscountRateStrategy(
+    address _ghoVariableDebtToken,
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:36]:
```suggestion
    address ghoVariableDebtToken,
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    address newDiscountRateStrategy
+  ) external onlyOwner {
+    IGhoVariableDebtToken(_ghoVariableDebtToken).updateDiscountRateStrategy(
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:39]:
```suggestion
    IGhoVariableDebtToken(ghoVariableDebtToken).updateDiscountRateStrategy(
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+   * @dev Constructor.
+   * @param owner The owner address of this contract
+   */
+  constructor(address owner) {
+    transferOwnership(owner);
+  }
+
+  /**
+   * @notice Updates the Discount Rate Strategy
+   * @param _ghoVariableDebtToken The address of GhoVariableDebtToken contract
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:32]:
```suggestion
   * @param ghoVariableDebtToken The address of GhoVariableDebtToken contract
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    address newDiscountRateStrategy
+  ) external onlyOwner {
+    IGhoVariableDebtToken(_ghoVariableDebtToken).updateDiscountRateStrategy(
+      newDiscountRateStrategy
+    );
+  }
+
+  /**
+   * @notice Updates the Discount Lock Period
+   * @param _ghoVariableDebtToken The address of GhoVariableDebtToken contract
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:46]:
```suggestion
   * @param ghoVariableDebtToken The address of GhoVariableDebtToken contract
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+   * @param newLockPeriod The new discount lock period (in seconds)
+   */
+  function updateDiscountLockPeriod(
+    address _ghoVariableDebtToken,
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:50]:
```suggestion
    address ghoVariableDebtToken,
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    uint256 newLockPeriod
+  ) external onlyOwner {
+    IGhoVariableDebtToken(_ghoVariableDebtToken).updateDiscountLockPeriod(newLockPeriod);
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:53]:
```suggestion
    IGhoVariableDebtToken(ghoVariableDebtToken).updateDiscountLockPeriod(newLockPeriod);
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  /**
+   * @notice Updates the ReserveInterestRateStrategy for Gho
+   * @param _poolConfigurator The address of PoolConfigurator contract
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:58]:
```suggestion
   * @param poolConfigurator The address of PoolConfigurator contract
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  function updateDiscountLockPeriod(
+    address _ghoVariableDebtToken,
+    uint256 newLockPeriod
+  ) external onlyOwner {
+    IGhoVariableDebtToken(_ghoVariableDebtToken).updateDiscountLockPeriod(newLockPeriod);
+  }
+
+  /**
+   * @notice Updates the ReserveInterestRateStrategy for Gho
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:57]:
```suggestion
   * @notice Updates the ReserveInterestRateStrategy
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+   * @param _poolConfigurator The address of PoolConfigurator contract
+   * @param asset The address of the GHO deployed contract
+   * @param newRateStrategyAddress The address of new RateStrategyAddress contract
+   */
+  function setReserveInterestRateStrategyAddress(
+    address _poolConfigurator,
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:63]:
```suggestion
    address poolConfigurator,
```

File: src/contracts/facilitators/aave/misc/GhoManager.sol [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    address asset,
+    address newRateStrategyAddress
+  ) external onlyOwner {
+    IPoolConfigurator(_poolConfigurator).setReserveInterestRateStrategyAddress(
```
Comment (miguelmtzinf) [inline src/contracts/facilitators/aave/misc/GhoManager.sol:67]:
```suggestion
    IPoolConfigurator(poolConfigurator).setReserveInterestRateStrategyAddress(
```

File: src/tasks/testnet-setup/07_add-gho-manager.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import { ghoEntityConfig } from '../../helpers/config';
+import { IGhoToken } from '../../../types';
+
+import {
+  getAaveProtocolDataProvider,
+  getProxyAdminBySlot,
+  STAKE_AAVE_PROXY,
+  ACL_MANAGER_ID,
+  getACLManager,
+} from '@aave/deploy-v3';
```
Comment (miguelmtzinf) [inline src/tasks/testnet-setup/07_add-gho-manager.ts:13]:
```suggestion
import { GhoManager } from '../../../types/src/contracts/facilitators/aave/misc/GhoManager';
import { task } from 'hardhat/config';
import { ghoEntityConfig } from '../../helpers/config';
import { getACLManager } from '@aave/deploy-v3';
```

**Review by miguelmtzinf — CHANGES_REQUESTED** [review]

**Pull request merged**